Templatize `_d_class_cast`

Albert24GG opened the pull request:
This PR introduces the templated version of `_d_class_cast`, cleaning up a bunch of code in `e2ir.d`.

I also had to add an additional check in `expressionsem.d` in case the casted expression is a variable, in which case its type needed to be stripped of `shared`. I am not sure if this is the cleanest or idiomatic way, but I am open to better solutions.

Albert24GG committed:
Templatize `_d_class_cast`

Replace `if` with `static if`

Use `Object` param for casting functions to avoid multiple instances

Revise comment

Remove unused function

Add unit test for `_d_class_cast`

Albert24GG commented:
> I'd like to have a comment stating something about this, that it doesn't need to be templated on the from type.
@rikkimax I forgot about that, thanks for bringing it up! As you mentioned, I ditched the `From` template parameter to reduce the number of instantiations, and that's ok since this compile time type was not really used anywhere.

kinke commented on druntime/src/core/internal/cast_.d:
```diff
@@ -63,6 +63,48 @@ private void* _d_paint_cast(To, From)(From o)
+ */
+private void* _d_class_cast(To)(Object o)
+{
+    debug(cast_) printf("_d_cast_cast(o = %p, c = '%.*s')\n", o, cast(int) c.name.length, c.name.ptr);
```
There's no `c` here.

Albert24GG committed:
Fix `_d_cast` instantiation with stripped `shared` arg attribute

Albert24GG committed:
Use cast on pointers in `atomicStore`

The cast performed in `atomicStore` is basically just an `Unqual` that
strips the `shared` attribute, so it can be safely done using a pointer
cast. This way, we don't need to skip `shared` checks in
`expressionsem.d`.

Albert24GG committed:
Use a template wrapper for `_d_class_cast`

The implementation of `_d_class_cast` does not directly use the template
parameter `To`, so it is a better idea to have it call
`_d_class_cast_impl` in order to avoid redundant code generations.

Albert24GG committed:
Use stronger attributes for the `Object` parameter

Albert24GG committed:
Remove prints

teodutu reviewed (approved):
Just address my comment about the extra brackets and keep the other one in mind for later. Otherwise, this is fine. GJ!

Albert24GG committed:
Remove stray brackets

Co-authored-by: Teodor Dutu <teodor.dutu@gmail.com>

Albert24GG commented on druntime/src/core/internal/cast_.d:
```diff
@@ -63,6 +57,50 @@ private void* _d_paint_cast(To, From)(From o)
+ * Returns:
+ *      null if `o` is null or `To` is not a subclass type of `o`. Otherwise, return `o`.
+ */
+private void* _d_class_cast(To)(const return scope Object o)
+{
+    return _d_class_cast_impl(o, typeid(To));
+}
```
Agreed, will do. Thanks!
